### DISCUSSION TITLE
How can I get the names of all of the circuits?

### DISCUSSION THREAD

**Discussion by ColtonIdle** [parnic/node-screenlogic#81] [Q&A] [answered]
So I've been digging into this library a lot. and it seems like with

you can get a list of basically everything attached to your system... + it's state (on/off) via `getEquipmentStateAsync.circuitArray`

What I want to do is actually get the names of all of those items from the above list. It seems like you can retrieve names in general via getAllCircuitNamesAsync, but that just gives you a list of names that are available to be **set** as a circuit/feature name. kinda like a dictionary.

Does anyone know how I'm suppose to map `getEquipmentStateAsync.circuitArray` to `getAllCircuitNamesAsync`
? Am I understanding that correctly?

**Comment by parnic** [comment]
It's been a while since I've messed with this, and my pool equipment is temporarily broken at the moment, but if I recall correctly the circuits in the array are ordered the same as the order of the names from `getCircuitNamesAsync`. So if you're trying to match a circuit to a name, you'd use the same index.

Does that help?

**Reply by ColtonIdle** [reply]
Hm. So the circuit array gives me about 20 items.

Let me try getCircuitNamesAsync. I was trying getAll*CircuitNamesAsync and was getting a huge alphabetical list of all possible names that you could assign to a circuit.

**Reply by parnic** [reply, marked as answer]
Ah, yeah, that method isn't documented and is only in there because we could (as in: the message exists and the system will respond to it).

You'll probably need v2.1.1 of the library for getCircuitNamesAsync as it was not working for me when I tried it while looking into this discussion.

**Reply by ColtonIdle** [reply]
> the message exists and the system will respond to it

Is there a way to see what messages exist?

**Reply by parnic** [reply]
> > the message exists and the system will respond to it
> 
> Is there a way to see what messages exist?

This entire project is the result of reverse engineering all these messages from various sources, heh. I wish there was. All we have is community documentation, Wireshark spying on any official apps, and reading decompiled app code.

**Reply by ColtonIdle** [reply]
Gotcha. I figured as much (ive tried my hand at wiresharking as well before i found this project). but yeah. was just curious if i missed the memo where all message names reside. lol

**Reply by ColtonIdle** [reply]
I was also playing around with the "Pentair SLConfig" app and that's actually where I noticed that the result from "getAll*CircuitNamesAsync" matched the list of choices in "Setup Circuits" in SLConfig app. Which is how I connected the dots that it's just a list of possible choices and not the actual names that were set.

FWIW. The list of names from getAll*CircuitNamesAsync ended up being like 50% of what the SLConfig app reports. Might make a bug report ticket just for the sake of trackign.

**Reply by ColtonIdle** [reply]
as always. thanks for the quick pair of eyes @parnic 
much appreciated

**Reply by ColtonIdle** [reply]
@parnic FWIW, I got this as my output now

circuitNames = {"senderId":0,"names":["USERNAME-01","USERNAME-02","USERNAME-03","USERNAME-04","USERNAME-05","USERNAME-06","USERNAME-07","USERNAME-08","USERNAME-09","USERNAME-10","","","","","","","","","",""]}

which doesn't seem right. I started peaking at the python and I see this

```
**************************
 ID  STATE  NAME
--------------------------
500    Off  Spa
501    Off  Pool Light
502    Off  BBQ Light
503    Off  Booster Pump
504    Off  Aux 4
505     On  Pool
506    Off  Aux 5
507    Off  Aux 6
508    Off  Aux 7
510    Off  Feature 1
511    Off  Pool High
512    Off  Feature 3
513    Off  Feature 4
514    Off  Feature 5
515    Off  Feature 6
516    Off  Feature 7
517    Off  Feature 8
519    Off  AuxEx
**************************
```

which is what I'm actually looking for. Do you know which command in the node library does this?

**Reply by parnic** [reply]
You should be able to get those names from `equipment.getControllerConfigAsync()`: https://github.com/parnic/node-screenlogic?tab=readme-ov-file#properties-6. I'm not sure where to get the list of IDs based at 500, though.

When you say "peeking at the python" - what Python were you looking at? Maybe there's a hint in there about what message they're sending.

**Reply by ColtonIdle** [reply]
Sorry. Peeking at the python lib == looking at this https://github.com/dieselrabbit/screenlogicpy

Will try getControllerConfigAsync... thanks!

**Reply by ColtonIdle** [reply]
Ah okay. So to get the output I posted from the python library I had to combine two calls in this library:

1. equipment.getControllerConfigAsync
2. equipment.getEquipmentStateAsync

and then I can combine the two to make the list from the python library. Hopefully that's "right" and theres not some easier way to do it. but that sounds like that should be it.

**Reply by parnic** [reply]
Sounds good. I can take a look at the python lib to see if there's a single message that gives both, but there should be very little harm in making two calls. The official app makes a _lot_ more than that frequently.

**Reply by ColtonIdle** [reply]
Cool. Yeah. I'm mostly just curious because the ordering isn't the same in both and so if I could save the round trip... why not
